Python/bltinmodule.c

```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+    return PyType_HasFeature(obj, Py_TPFLAGS_IMMUTABLETYPE) || !PyType_HasFeature(obj, Py_TPFLAGS_HEAPTYPE);
+}
+
+PyObject* add_elements_fast(PyObject* seq, stack* frontier) {
+    Py_ssize_t size;
+    _PyObject_ASSERT(seq, PyList_Check(seq) || PyTuple_Check(seq));
+    size = PySequence_Fast_GET_SIZE(seq);
+    _Py_VPYDBG("list/tuple: pushing %ld [i: element] elements\n", size);
+    for(Py_ssize_t i = 0; i < size; i++){
+        PyObject* element = PySequence_Fast_GET_ITEM(seq, i); // element.rc = x
+        _Py_VPYDBG("[%ld: ", i);
+        _Py_VPYDBGPRINT(element);
+        _Py_VPYDBG("]\n");
+        if(!_Py_IsImmutable(element)){
+            Py_INCREF(element); // element.rc = x + 1
+            if(stack_push(frontier, element)){
+                Py_DECREF(element);
+                return PyErr_NoMemory();
+            }
+        }
+    }
+    _Py_VPYDBG("\n");
+    Py_RETURN_NONE;
+}
+
+PyObject* add_elements(PyObject* seq, stack* frontier) {
+    Py_ssize_t size;
+    _PyObject_ASSERT(seq, PySequence_Check(seq));
+    size = PySequence_Size(seq);
+    _Py_VPYDBG("list/tuple: pushing %ld [i: element] elements\n", size);
+    for(Py_ssize_t i = 0; i < size; i++){
+        PyObject* element = PySequence_GetItem(seq, i); // element.rc = x + 1
+        _Py_VPYDBG("[%ld: ", i);
+        _Py_VPYDBGPRINT(element);
+        _Py_VPYDBG("]\n");
+        if(!_Py_IsImmutable(element)){
+            if(stack_push(frontier, element)){
+                Py_DECREF(element); // element.rc = x
+                return PyErr_NoMemory();
+            }
+        }else{
+            Py_DECREF(element); // element.rc = x
+        }
+    }
+    _Py_VPYDBG("\n");
+    Py_RETURN_NONE;
+}
+
+PyObject* walk_sequence(PyObject* seq, stack* frontier) {
+    if(PyList_Check(seq) || PyTuple_Check(seq)){
+        return add_elements_fast(seq, frontier);
+    }else if(PySequence_Check(seq)){
+        return add_elements(seq, frontier);
+    }
+
+    Py_RETURN_NONE;
+}
+
+PyObject* walk_mapping(PyObject* map, stack* frontier) {
+    if(PyMapping_Check(map) && (PyDict_CheckExact(map) || PyObject_HasAttrString(map, "keys"))){
+        PyObject* keys = PyMapping_Keys(map);
+        Py_ssize_t size = PyList_Size(keys);
+        _Py_VPYDBG("mapping: pushing %ld <key: value> pairs\n", size);
+        for(Py_ssize_t i = 0; i < size; i++){
+            PyObject* key = PyList_GET_ITEM(keys, i); // key.rc = x
+            _Py_VPYDBG("<");
+            _Py_VPYDBGPRINT(key);
+            if(!_Py_IsImmutable(key)){
+                Py_INCREF(key); // key.rc = x + 1
+                if(stack_push(frontier, key)){
+                    Py_DECREF(key); // key.rc = x
+                    return PyErr_NoMemory();
+                }
+            }
+
+            PyObject* value = PyObject_GetItem(map, key); // value.rc = x + 1
+            _Py_VPYDBG(": ");
+            _Py_VPYDBGPRINT(value);
+            if(!_Py_IsImmutable(value)){
+                if(stack_push(frontier, value)){
```

Comment:
I assume `PyObject_GetItem` returns a RC on the object it returns.  So I think we are counting twice in the successful case.